File: swift/Solutions/Day14.swift
//
//  Day14.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/12/23.
//

import Foundation

struct Day14: Day {
    func part1Solution(input: [String]) -> String {
        var grid = input.filter({!$0.isEmpty}) .map { Array($0) }
        
        func trickleUp(_ x: Int, _ y: Int){
            if x == 0 { return }
            if grid[x][y] == "O", grid[x-1][y] == "." {
                grid[x][y] = "."
                grid[x-1][y] = "O"
                trickleUp(x-1, y)
            }
        }
        
        for i in 0..<grid.count {
            for j in 0..<grid[0].count {
                if grid[i][j] == "O" {
                    trickleUp(i, j)
                }
            }
        }
        
        var answer = 0
        for i in 0..<grid.count {
            for j in 0..<grid[0].count {
                if grid[i][j] == "O" {
                    answer += (input.filter({!$0.isEmpty}).count - i)
                }
            }
        }
        
        return "\(answer)"
    }
    
    func part2Solution(input: [String]) -> String {
        var grid = input.filter({!$0.isEmpty}) .map { Array($0) }
        
        func trickleUp(_ x: Int, _ y: Int){
            if x == 0 { return }
            if grid[x][y] == "O", grid[x-1][y] == "." {
                grid[x][y] = "."
                grid[x-1][y] = "O"
                trickleUp(x-1, y)
            }
        }
        func trickleLeft(_ x: Int, _ y: Int){
            if y == 0 { return }
            if grid[x][y] == "O", grid[x][y-1] == "." {
                grid[x][y] = "."
                grid[x][y-1] = "O"
                trickleLeft(x, y-1)
            }
        }
        func reaganomics(_ x: Int, _ y: Int){
            if x == grid.count-1 { return }
            if grid[x][y] == "O", grid[x+1][y] == "." {
                grid[x][y] = "."
                grid[x+1][y] = "O"
                reaganomics(x+1, y)
            }
        }
        func trickleRight(_ x: Int, _ y: Int){
            if y == grid[0].count-1 { return }
            if grid[x][y] == "O", grid[x][y+1] == "." {
                grid[x][y] = "."
                grid[x][y+1] = "O"
                trickleRight(x, y+1)
            }
        }
        
        var cache = [String: Int]()
        
        func printGrid(){
            for line in grid {
                print(String(line))
            }
            
            print(" ")
        }
        
        for iteration in 0..<cycleNum {
            for i in 0..<grid.count {
                for j in 0..<grid[0].count {
                    if grid[i][j] == "O" {
                        trickleUp(i, j)
                    }
                }
            }
            for j in 0..<grid.count {
                for i in 0..<grid[0].count {
                    if grid[i][j] == "O" {
                        trickleLeft(i, j)
                    }
                }
            }
            for i in (0..<grid.count).reversed(){
                for j in 0..<grid[0].count {
                    if grid[i][j] == "O" {
                        reaganomics(i, j)
                    }
                }
            }
            for j in (0..<grid.count).reversed() {
                for i in 0..<grid[0].count {
                    if grid[i][j] == "O" {
                        trickleRight(i, j)
                    }
                }
            }
            let endState = String(grid.joined(separator: ""))
            if let prevIteration = cache[endState], (cycleNum - prevIteration ) % ((iteration + 1) - prevIteration) == 0 {
                break
            }
            cache[endState] = iteration + 1
        }
        
        var answer = 0
        for i in 0..<grid.count {
            for j in 0..<grid[0].count {
                if grid[i][j] == "O" {
                    answer += (grid.count - i)
                }
            }
        }
        
        return "\(answer)"
    }
    
    let cycleNum = 1000000000
}

File: kotlin/src/main/kotlin/solutions/Day14.kt
package solutions

class Day14: Day {
    private lateinit var grid: MutableList<CharArray>
    private val trickles
        get() = listOf(
            Trickle(
                outerRange = 0..<grid.size,
                innerRange = 0..<grid.first().size,
                trickle = { x, y -> Triple(x - 1, y, x == 0) }
            ),
            Trickle(
                outerRange = 0..<grid.first().size,
                innerRange = 0..<grid.size,
                trickle = { x, y -> Triple(x, y - 1, y == 0) }
            ),
            Trickle(
                outerRange = (grid.size - 1) downTo 0,
                innerRange = 0..<grid.first().size,
                trickle = { x, y -> Triple(x + 1, y, x == grid.size - 1) }
            ),
            Trickle(
                outerRange = (grid.first().size - 1) downTo 0,
                innerRange = 0..<grid.size,
                trickle = { x, y -> Triple(x, y + 1, y == grid[0].size - 1) }
            ),
        )


    private fun trickle(x: Int, y: Int, direction: (Int, Int) -> Triple<Int, Int, Boolean>){
        val (nextX, nextY, edgeReached) = direction(x, y)
        if (edgeReached) return
        if (grid[x][y] == 'O' && grid[nextX][nextY] == '.') {
            grid[x][y] = '.'
            grid[nextX][nextY] = 'O'
            trickle(nextX, nextY ,direction)
        }
    }
    override fun part1Solution(input: List<String>): String {
        grid = input.mapNotNull { if(it.isNotEmpty()) it.toCharArray() else null }.toMutableList()
        for (i in trickles.first().outerRange){
            for (j in trickles.first().innerRange){
                trickle(i, j, trickles.first().trickle)
            }
        }
        var answer = 0
        grid.forEachIndexed { index, chars ->
            chars.forEach {
                if (it == 'O') answer += (grid.size - index)
            }
        }
        return answer.toString()
    }

    override fun part2Solution(input: List<String>): String {
        val cycles = 1_000_000_000L
        grid = input.mapNotNull { if(it.isNotEmpty()) it.toCharArray() else null }.toMutableList()
        val cache = mutableMapOf<String, Long>()
        for (iteration in 0 until cycles) {
            trickles.forEachIndexed { index, direction ->
                for (i in direction.outerRange) {
                    for (j in direction.innerRange) {
                        trickle(
                            if (index % 2 == 0) i else j,
                            if (index % 2 == 0) j else i,
                            direction.trickle
                        )
                    }
                }
            }
            val gridState = grid.joinToString(separator = "") { it.joinToString(separator = "") }
            val prev = cache[gridState]
            if(prev != null){
                if ((cycles - prev ) % ((iteration + 1) - prev) == 0L){
                    break
                }
            }
            cache[gridState] = (iteration + 1)
        }
        var answer = 0
        grid.forEachIndexed { index, chars ->
            chars.forEach {
                if (it == 'O') answer += (grid.size - index)
            }
        }
        return answer.toString()
    }

    data class Trickle(
        val outerRange: IntProgression,
        val innerRange: IntProgression,
        /// Returns (nextX, nextY, edgeReached)
        val trickle: (Int, Int) -> Triple<Int, Int, Boolean>
    )
}